server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java

```diff
@@ -342,6 +381,28 @@ public void canReturnNullResponseIfMatchNoDocs(boolean value) {
         this.canReturnNullResponseIfMatchNoDocs = value;
     }
 
+    /**
+     * Returns <code>true</code> if the request needs to create a long-lived context.
+     */
+    public boolean shouldCreatePersistentReader() {
+        return scroll != null  || (keepAlive != null && readerId == null);
```

Comment:
Should it be `readerId != null`?

Reply:
This was needed when the search context was created on the first search. We have a dedicated API to create search context now so this condition is never met (readerId cannot be null). I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/d654b263970c570677c951fbdfb12018b9311019.